blocks/hooks/notices/index.js

```diff
@@ -0,0 +1,86 @@
+/**
+ * External dependencies
+ */
+import uuid from 'uuid/v4';
+
+/**
+ * WordPress dependencies
+ */
+import { Component, getWrapperDisplayName } from '@wordpress/element';
+import { addFilter } from '@wordpress/hooks';
+import { NoticeList } from '@wordpress/components';
+
+/**
+ * Internal dependencies
+ */
+import { hasBlockSupport } from '../../api';
+import './editor.scss';
+
+/**
+ * Override the default edit UI to include notices if supported.
+ *
+ * @param   {function|Component} BlockEdit Original component.
+ * @returns {Component}                    Wrapped component.
+ */
+export function withNotices( BlockEdit ) {
+	class WrappedBlockEdit extends Component {
+		/** @inheritdoc */
+		constructor() {
+			super( ...arguments );
+
+			this.addNotice = this.addNotice.bind( this );
+			this.removeNotice = this.removeNotice.bind( this );
+
+			this.state = {
+				notices: [],
+			};
+		}
+
+		/**
+		* Function passed down to blocks as a prop that adds a new notice.
+		*
+		* @param {Object} notice  Notice to add.
+		*/
+		addNotice( notice ) {
+			const noticeToAdd = notice.id ? notice : { ...notice, id: uuid() };
+			this.setState( state => ( {
+				notices: [ ...state.notices, noticeToAdd ],
+			} ) );
+		}
+
+		/**
+		* Removes a notice by id.
+		*
+		* @param {string} id  Id of the notice to remove.
+		*/
+		removeNotice( id ) {
```

Comment:
What happens when a block with a notice disappears? Does it just stay around here in the state? This is why I think it would be beneficial for the state to be local, not global.

Reply:
In other words, why do we need to keep a global list?

Reply:
Notices is a HoC around blocks, it would be the same as doing something like withNotices( GalleryBlock ); that adds notices property to gallery block. I can easily switch to this version. The only disadvantage is that blocks would need to be React components to use it and the current approach also allows blocks to take advantage even if their edit method is a function. When a block with notices is removed, the HoC around it will also be removed and its state is deleted. 
Similar to happens when we use withAPIData the HoC also has its own state when a block it wraps is removed its own state is also deleted.